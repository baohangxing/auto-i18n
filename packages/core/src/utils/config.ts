import fs from 'fs'
import path from 'path'
import consola from 'consola'
import type { AutoConfig, Config, LangJson } from '../types'

const configNames = ['auto.config.ts', 'auto.config.js']

const defaultAutoConfig: Required<AutoConfig> = {
  localesJsonDirs: './src/lang/locales',
  locales: ['zh-cn', 'ja-jp', 'ko-kr'],
  baseLocale: 'zh-cn',
  transLacaleWord(word: string, locale: string, toLocale: string) {
    return Promise.resolve(`[${toLocale.toUpperCase()}]${word}`)
  },
  includes: [],
  outputFileDir: './',
}

const readAutoConfig = async (): Promise<AutoConfig> => {
  for (const configName of configNames) {
    const configFilePath = path.join(process.cwd(), configName)
    if (!fs.existsSync(configFilePath)) {
      const res = await import(configFilePath)
      console.log(res)
      return res
    }
  }
  consola.error(`Pleace add ${configNames.join(' or ')} file in your project(${process.cwd()}) and export default AutoConfig`)
  process.exit(1)
}

const getAutoConfig = (() => {
  let autoConfig: Required<AutoConfig>
  (async () => {
    autoConfig = Object.assign({}, defaultAutoConfig, await readAutoConfig())
    return () => {
      return autoConfig
    }
  })()
  return () => autoConfig
})()

const getJsonPath = (autoConfig: AutoConfig): {
  baseLangJson: LangJson
  otherLangJsons: LangJson[]
} => {
  const dirs = typeof autoConfig.localesJsonDirs === 'string'
    ? [autoConfig.localesJsonDirs]
    : autoConfig.localesJsonDirs

  let otherLangJsons: LangJson[] = []

  const files: string[] = []

  for (const dir of dirs) {
    if (fs.existsSync(dir) && fs.statSync(dir).isDirectory()) {
      const children = fs.readdirSync(dir)
      for (const x of children) {
        const childPath = path.join(dir, x)
        if (fs.statSync(childPath).isFile())
          files.push(childPath)
      }
    }
  }

  for (const x of autoConfig.locales) {
    let findFlag = false
    for (const file of files) {
      if (path.parse(file).name === x) {
        findFlag = true
        otherLangJsons.push(
          {
            name: x,
            path: file,
          },
        )
        break
      }
    }
    if (!findFlag) {
      consola.error(`No json file(${x}.json) in AutoConfig.localesJsonDirs: ${autoConfig.localesJsonDirs}`)
      process.exit(1)
    }
  }
  let baseLangJson: LangJson = { name: autoConfig.baseLocale, path: '' }
  if (autoConfig.baseLocale.includes(autoConfig.baseLocale)) {
    baseLangJson = otherLangJsons.find(x => x.name === autoConfig.baseLocale) as LangJson
    otherLangJsons = otherLangJsons.filter(x => x.name !== autoConfig.baseLocale)
  }
  else {
    for (const file of files) {
      if (path.parse(file).name === autoConfig.baseLocale) {
        baseLangJson.path = file
        break
      }
    }
    if (!baseLangJson.path) {
      consola.error(`No json file(${autoConfig.baseLocale}.json) in AutoConfig.localesJsonDirs: ${autoConfig.localesJsonDirs}`)
      process.exit(1)
    }
  }

  return {
    baseLangJson,
    otherLangJsons,
  }
}

const config: Config = (() => {
  const autoConfig = getAutoConfig()

  const { baseLangJson, otherLangJsons } = getJsonPath(autoConfig)
  return {
    baseLangJson,
    otherLangJsons,
  }
})()

export { getAutoConfig, config }
