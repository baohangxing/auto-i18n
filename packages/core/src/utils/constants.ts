export const LOG_PREFIX = 'yo-auto-i18n'

export const CLI_CONFIG_NAME = 'auto'

export const KEY_SYMBOL_IN_XLSX = 'key'

export const XLSX_NAME = 'all_locales'
