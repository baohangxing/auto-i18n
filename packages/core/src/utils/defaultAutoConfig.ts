import type { AutoConfig, I18nCallRule } from '../type'

const getCommonI18nCallRule = (): I18nCallRule => {
  return {
    caller: '',
    functionName: 't',
    definedDeclaration: 'const { t } = useI18n();',
    importDeclaration: 'import { useI18n } from "vue-i18n";',
  }
}

const defaultAutoConfig: AutoConfig = {
  localesJsonDirs: ['./src/lang/locales'],
  locales: ['zh-cn', 'ja-jp', 'ko-kr'],
  baseLocale: 'zh-cn',
  untransSymbol: (locale: string) => {
    return `[${locale.toUpperCase()}]`
  },
  includes: [],
  outputFileDir: './',
  exclude: [],
  i18nCallRules: {
    js: getCommonI18nCallRule(),
    ts: getCommonI18nCallRule(),
    cjs: getCommonI18nCallRule(),
    mjs: getCommonI18nCallRule(),
    jsx: getCommonI18nCallRule(),
    tsx: getCommonI18nCallRule(),
    vue: getCommonI18nCallRule(),
  },
}

export default defaultAutoConfig
