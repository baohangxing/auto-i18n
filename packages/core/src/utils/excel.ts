import fs from 'fs'
import type { WorkSheet } from 'node-xlsx'
import xlsx from 'node-xlsx'
import consola from 'consola'

const writeXlsxFile = (
  firstRows: string[][],
  sheetNameArr: string[],
  sheetsDatas: string[][][],
  fileWritePath: string,
) => {
  const data: WorkSheet<string>[] = []

  for (let i = 0; i < sheetNameArr.length; i++) {
    data.push({
      name: sheetNameArr[i],
      data: [firstRows[i], ...sheetsDatas[i]],
      options: {},
    })
  }

  const buffer = xlsx.build(data)

  fs.writeFile(fileWritePath, buffer, (err) => {
    if (err)
      consola.error(err)
    consola.info(`Write to xls has finished: ${fileWritePath}`)
  })
}

const readXlsxFile = (filePath: string) => {
  if (fs.existsSync(filePath)) {
    const obj = xlsx.parse(filePath)
    return obj as {
      name: string
      data: string[][]
    }[]
  }
  else {
    consola.error(`not exists ${filePath}`)
  }
}

export { writeXlsxFile, readXlsxFile }
