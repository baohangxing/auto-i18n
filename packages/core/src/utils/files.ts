import fs from 'fs'
import path from 'path'

const writeFileSyncForce = (
  file: string,
  data: string | NodeJS.ArrayBufferView,
  options?: fs.WriteFileOptions | undefined) => {
  if (!fs.existsSync(path.join(file, '..')))
    fs.mkdirSync(path.join(file, '..'), { recursive: true })
  fs.writeFileSync(file, data, options)
}

export { writeFileSyncForce }
