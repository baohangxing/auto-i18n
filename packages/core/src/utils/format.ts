import fs from 'fs'
import path from 'path'
import { ESLint } from 'eslint'

/**
 * check ESlint Config (.eslintrc.* file) exist
 * @param recursive find from process.cwd() to root recursively
 * @returns
 */
const checkEslintConfigExist = (cwd = process.cwd(), recursive = false): [boolean, string] => {
  const maybeNames = ['.eslintrc.js', '.eslintrc.cjs',
    '.eslintrc.yaml', '.eslintrc', '.eslintrc.yml', '.eslintrc.json']

  const checkHelp = (checkPath: string): boolean => {
    for (const name of maybeNames) {
      const eslintrcPath = path.join(checkPath, name)
      if (fs.existsSync(eslintrcPath))
        return true
    }
    return false
  }

  if (recursive) {
    let startPath = cwd
    let nextPath = path.join(startPath, '..')

    if (checkHelp(startPath))
      return [true, startPath]

    while (startPath !== nextPath) {
      if (checkHelp(nextPath))
        return [true, nextPath]
      startPath = nextPath
      nextPath = path.join(startPath, '..')
    }

    return [false, '']
  }
  else {
    return checkHelp(cwd) ? [true, cwd] : [false, '']
  }
}

const initEslint = () => {
  const [useEslintrc, path] = checkEslintConfigExist(process.cwd(), true)
  const eslint = new ESLint({
    cwd: useEslintrc ? path : process.cwd(),
    useEslintrc,
    fix: true,
  })
  return eslint
}

const lintFiles = async (paths: string | string[]) => {
  const eslint = initEslint()

  const filePaths = ([] as string[]).concat(paths)

  for (let i = 0; i < filePaths.length; i++) {
    try {
      const results = await eslint.lintFiles([filePaths[i]])
      const result = results[0]
      if (result && result.output)
        fs.writeFileSync(filePaths[i], result.output, 'utf8')
    }
    catch (error) {
      console.error(error)
    }
  }
}

const lintTextByCreateFile = async (testCode: string, fileExt = 'ts'): Promise<string> => {
  const tmpPath = path.join(process.cwd(), `___eslint_${new Date().getTime()}.${fileExt}`)
  fs.writeFileSync(tmpPath, testCode, 'utf8')
  let result = testCode
  try {
    await lintFiles([tmpPath])
    result = fs.readFileSync(tmpPath, 'utf8')
    fs.rmSync(tmpPath)
  }
  catch (error) {
    fs.rmSync(tmpPath)
    throw error
  }

  return result
}

const lintText = async (testCode: string, fileExt = 'ts'): Promise<string> => {
  const eslint = initEslint()
  const result = await eslint.lintText(testCode)
  if (result?.[0].output === undefined)
    return await lintTextByCreateFile(testCode, fileExt)

  return result?.[0].output
}

export { checkEslintConfigExist, lintFiles, lintText }
