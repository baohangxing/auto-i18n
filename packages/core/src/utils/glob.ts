// https://github.com/mrmlnc/fast-glob

import path from 'path'
import fs from 'fs'
import fg from 'fast-glob'
import type { Options } from 'fast-glob'
import log from './log'

const fgOptions: Options = {
  absolute: true,
  ignore: ['**/node_modules/**', '**/*.d.ts'],
}

/** get absolute paths with the glob patterns */
const fgSync = (patterns: string[] | string): string[] => {
  const source = ([] as string[]).concat(patterns)
  return fg.sync(source, fgOptions)
}

/** wheather the patterns match the absolute path or not */
const checkInPatterns = (path: string, patterns: string[]): boolean => {
  const paths = fgSync(patterns)
  return paths.includes(path)
}

/** get absolute paths by a dir or file path */
const getPaths = (dirOrFile: string): string[] => {
  const dirOrFilePath = path.join(process.cwd(), dirOrFile)

  if (!fs.existsSync(dirOrFilePath)) {
    log.error(`${dirOrFilePath} not exists`)
    return []
  }

  if (fs.statSync(dirOrFilePath).isDirectory()) {
    const files: string[] = []
    const dirs = [dirOrFilePath]

    while (dirs.length) {
      const dir: string = dirs.pop() as string
      if (!fs.existsSync(dir))
        continue
      if (fs.statSync(dir).isDirectory()) {
        const children = fs.readdirSync(dir)
        for (const x of children) {
          const childPath = path.join(dir, x)
          if (fs.statSync(childPath).isFile())
            files.push(childPath)
          else if (!childPath.includes('node_modules'))
            dirs.push(childPath)
        }
      }
      else {
        files.push(dir)
      }
    }
    return files
  }
  else {
    return [dirOrFilePath]
  }
}

export { fgSync, getPaths, checkInPatterns }
