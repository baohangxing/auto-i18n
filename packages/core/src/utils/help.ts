import path from 'path'
import fs from 'fs'
import { getAutoConfig } from '../config/config'
import log from './log'
import { checkInPatterns } from './glob'
import { format } from './format'

const includeChinese = (code: string) => {
  return /[\u4E00-\u9FFF]/g.test(code)
}

const createFileName = (fileName = '') => {
  const data = new Date()
  const name = `${fileName}_${
    data.getMonth() + 1
  }_${data.getDate()} ${data.getHours()}_${
    data.getMinutes() <= 9 ? `0${data.getMinutes()}` : data.getMinutes()
  }_${
    data.getSeconds() <= 9 ? `0${data.getSeconds()}` : data.getSeconds()
  }`

  return name
}

const lexicalComparator = (a: string, b: string) => {
  a = a.toString()
  b = b.toString()
  const n = Math.max(a.length, b.length)
  for (let i = 0; i < n; i++) {
    if (a[i] === undefined)
      return -1

    if (b[i] === undefined)
      return 1

    if (a[i] !== b[i])
      return a.charCodeAt(i) - b.charCodeAt(i)
  }
  return -1
}

/**
 * sort Object by lexical, retren a new Object
 * @param obj
 * @returns
 */
const sortObjectKey = (obj: any): any => {
  if (!obj)
    return {}
  const a: any = {}

  const keys = Object.keys(obj).sort(lexicalComparator)

  for (const key of keys) {
    if (typeof obj[key] === 'object')
      a[key] = sortObjectKey(obj[key])
    else
      a[key] = obj[key]
  }
  return a
}

const writeToJsonFile = async (
  writeToPath: string,
  name: string,
  obj: object,
) => {
  const jsonPath = path.join(writeToPath, `${name}.json`)
  const autoConfig = getAutoConfig()
  fs.writeFileSync(jsonPath, JSON.stringify(sortObjectKey(obj), undefined, 2), 'utf8')

  if (autoConfig.autoFormat && checkInPatterns(jsonPath, autoConfig.autoFormatRules))
    await format(jsonPath)
}

const getValueByKey = (obj: any, key: string) => {
  if (key === '')
    return obj
  const paths = key.split('.').filter(x => x !== '')

  let temp = obj

  if (temp[key] && typeof temp[key] === 'string')
    return temp[key]

  while (paths.length) {
    const path = paths.shift()

    if (path && temp[path] !== undefined)
      temp = temp[path]
    else
      return undefined
  }
  return temp
}

const setValueByKey = (obj: any, key: string, val: string) => {
  const paths = key.split('.').filter(x => x !== '')

  if (paths.length === 0)
    return

  let temp = obj
  while (paths.length > 1) {
    const path = paths.shift()

    if (path) {
      if (temp[path] === undefined)
        temp[path] = {}

      temp = temp[path]
      if (typeof temp !== 'object') {
        log.error(`can not set ${val} to the ${key}`)
        return
      }
    }
  }
  temp[paths[0]] = val
}

const getKeys = (obj: any) => {
  const _getKeys = (obj: any, _before = '') => {
    const res: string[] = []

    for (const k of Object.keys(obj)) {
      if (typeof obj[k] === 'string')
        res.push(_before === '' ? k : `${_before}.${k}`)
      else if (typeof obj[k] === 'object' && !(Array.isArray(obj[k])))
        res.push(..._getKeys(obj[k], _before === '' ? k : `${_before}.${k}`))
      else
        log.error(`the i18n JSON value only support string (${k}: ${obj[k]})`)
    }
    return res
  }
  return _getKeys(obj, '')
}

const escapeQuotes = (value: string): string => {
  return value.replaceAll('\'', '\\\'').replaceAll('"', '\\"')
}

export {
  writeToJsonFile,
  lexicalComparator,
  sortObjectKey,
  getValueByKey,
  setValueByKey,
  getKeys,
  createFileName,
  includeChinese,
  escapeQuotes,
}
