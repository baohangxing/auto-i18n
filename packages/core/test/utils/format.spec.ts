import path from 'path'
import fs from 'fs'
import { describe, expect, it } from 'vitest'
import { checkEslintConfigExist, lintFiles, lintText } from '../../src/utils/format'
import { trimRootPath } from '../_helps/utils'

describe('#checkEslintConfigExist', () => {
  it('should return false', async () => {
    const [reslut, configPath] = checkEslintConfigExist(process.cwd(), false)
    expect(reslut).toEqual(false)
    expect(configPath).toEqual('')
  })

  it('should return true', async () => {
    const root = path.join(process.cwd(), '..', '..')
    const [reslut, configPath] = checkEslintConfigExist(process.cwd(), true)
    expect(reslut).toEqual(true)
    expect(trimRootPath([configPath], root)).toMatchSnapshot()
  })
})

describe('#format', () => {
  it('should format file content', async () => {
    const filePath = path.join(__dirname, '..', '..', '..', '..', 'scripts', 'release.ts')

    await lintFiles(filePath)
    const res = fs.readFileSync(filePath, 'utf-8')

    expect(res).toMatchSnapshot()
  })

  it('should lint text', async () => {
    const content = `/**
    * sort Object by lexical recursively, retren a new Object.
    * not support array value
    * @param obj
    * @returns
    */
    const  sortObjectKey = (obj: any): any => {
    if (!obj)   return {}
    const a: any = {};const  keys =    Object.keys( obj ).sort(lexicalComparator)
    for (const key of keys) { if (typeof obj[key] === 'object')
        a[key] = sortObjectKey(obj[key])
      else
        a[key] = obj[key]
    }
      return a
    }`

    const reslut = await lintText(content)
    expect(reslut).toMatchSnapshot()
    expect(reslut).not.toEqual(content)
  })
})
