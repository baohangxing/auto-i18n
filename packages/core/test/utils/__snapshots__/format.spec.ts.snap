// Vitest Snapshot v1

exports[`#checkEslintConfigExist > should return true 1`] = `
[
  "",
]
`;

exports[`#format > should lint text 1`] = `
"/**
    * sort Object by lexical recursively, retren a new Object.
    * not support array value
    * @param obj
    * @returns
    */
const sortObjectKey = (obj: any): any => {
  if (!obj)
    return {}
  const a: any = {}; const keys = Object.keys(obj).sort()
  for (const key of keys) {
    if (typeof obj[key] === 'object')
      a[key] = sortObjectKey(obj[key])
    else
      a[key] = obj[key]
  }
  return a
}
export { sortObjectKey }
"
`;
